### PR TITLE
Add benchmarking for tests

### PR DESCRIPTION
**Pull request by Glitch18**
Adds an automation to `make test_yul` that benchmarks the number of steps per function invocation for each contract in a markdown called `stats.md`

I've switched the number of workers in the MakeFile to 1 for the yul tests. This is to prevent concurrent writes to `stats.json` through `stepsInFunction()`. Will find a fix for this. 

Will be adding support for benchmarking contract sizes.

### CODE REVIEW

**Comment by Glitch18** [comment]
This creates sort of an API that can be used in benchmarking tests for logging the results.
```
stepsInFunction(contractName, functionName, resultObject)
bytecodeDetails(contractName, bytecode)
sizeOfFile(contractName, file)
```

To see an example of these functions, please have a look at `tests/benchmark`

**Comment by Glitch18** [comment]
Regarding `WETH10`:

 - test for `receive()` is commented out because warp does not support this special function yet.
 - tests for `withdraw` and related functions are commented out because they were failing for the compiled cairo with an error saying address out of range. I believe this is because of the `(bool success, ) = to.call{value: value}("");` statements in these functions.

**Comment by Glitch18** [comment]
Regarding `IntegralMath`:

**Handwritten Cairo**:

 - Test for `ceilSqrt` fails due to `uint256_exp` with a Starknet exception saying "not yet implemented"
 - Similarly, tests for `roundDiv`, `mulDivF`, 'mulDivC', `mul512' fail with the same Starknet exception because of functions like `uint256_mod` and others.
 
 **Compiled Cairo**: 

Strangely, for the compiled cairo, the call to `compile_starknet_files` fails with a PreProcessorError while importing files.

**Comment by Glitch18** [comment]
Configuration for Github Action:

The plan for this PR is to implement a Github Action that will automatically use the benchmarking utilities and generate a markdown file that will be pushed to a separate repository.

As of now, the action is configured to target a receiving repository with the name "NethermindEth/warp-benchmark"

File: Makefile [hunk 1]
```diff
@@ -25,10 +25,16 @@ test_bats: warp $(BATS_FILES) tests/cli/*.bats
 test_yul: warp
 	python -m pytest scripts/yul/transpile_test.py -v --tb=short --workers=auto $(ARGS)
 	python -m pytest scripts/yul/compilation_test.py -v --tb=short --workers=auto $(ARGS)
-	python -m pytest tests/yul/ -v --tb=short --workers=auto $(ARGS)
+	python -m pytest tests/yul/ -v --tb=short --workers=1 $(ARGS)
```
[the remark this answers is not in the document]
Comment (murcake) [reply Makefile:28]: Yeah, I agree with the point about individual benchmark files

File: tests/benchmark/BaseJumpRateModelV2_compiled.cairo [hunk 2]
```diff
@@ -0,0 +1,780 @@
+%lang starknet
```
Comment (murcake) [inline tests/benchmark/BaseJumpRateModelV2_compiled.cairo:1]: Since it is a generated file, do we need it in the repo?

**Comment by murcake** [comment]
I'm not sure why the CI isn't starting, maybe it's because you are doing a PR from a separate repo? Could you please create a branch in this repo and make a PR from here?

**Comment by murcake** [comment]
> As of now, the action is configured to target a receiving repository with the name "NethermindEth/warp-benchmark"

I can't access that repository

**Comment by Glitch18** [comment]
> I'm not sure why the CI isn't starting, maybe it's because you are doing a PR from a separate repo? Could you please create a branch in this repo and make a PR from here?

Ah I see the error. It says `a step cannot have both the 'uses' and 'run' keys` I'll fix this and update my PR.

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (Glitch18) [reply Makefile:28]: Yes. The individual benchmark files make sense. Will implement this. Thanks @murcake and @swapnilraj

File: tests/benchmark/BaseJumpRateModelV2_compiled.cairo (hunk 2 above)
Comment (Glitch18) [reply tests/benchmark/BaseJumpRateModelV2_compiled.cairo:1]: Agreed. WIll remove it. Thanks for pointing this out.

**Comment by Glitch18** [comment]
> I can't access that repository

So after my previous comment mentioning "NethermindEth/warp-benchmark", I updated the action to pick up the target repository name from a variable in github secrets. We can set this variable once we've created the benchmark repository.

**Comment by swapnilraj** [comment]
The Github UI is confusing me a bit, but did you mean to delete `warp/bin/linuxkudu`

**Comment by Glitch18** [comment]
> The Github UI is confusing me a bit, but did you mean to delete `warp/bin/linuxkudu`

Now that I look at it, the Github UI does seem confusing. But I haven't deleted the binary. I replaced it with a newer version of it that I got from artifacts of the kudu repository.

**Comment by murcake** [comment]
image
Can you please fix the CI issue?

**Comment by Glitch18** [comment]
The CI issue above has been fixed. 

As of now, everything for this PR is complete except for a minor issue where running the following test errors out:

`python -m pytest scripts/yul/transpile_test.py -v --tb=short --workers=auto`

All other tests work fine. I'm looking into this right now and will upload a fix soon.

**Comment by Glitch18** [comment]
I've figured out why `python -m pytest scripts/yul/transpile_test.py -v --tb=short --workers=auto` is failing with my changes. This is because I'm using an updated binary of kudu.

I need to use the newer kudu as it translates `stop` to `revert(0,0)` without which the WETH10 contract fails with the following error upon encountering stop:

```
E               starkware.cairo.lang.compiler.preprocessor.preprocessor_error.PreprocessorError: /home/glitch/Github/warp/tests/benchmark/WETH10_compiled.cairo:3016:9: Unknown identifier 'stop'.
E                       stop()
E                       ^**^
```

**Comment by swapnilraj** [comment]
> > The Github UI is confusing me a bit, but did you mean to delete `warp/bin/linuxkudu`
> 
> Now that I look at it, the Github UI does seem confusing. But I haven't deleted the binary. I replaced it with a newer version of it that I got from artifacts of the kudu repository.

Can you update the macos kudus as well, we should keep them in sync.

**Comment by murcake** [comment]
> > > The Github UI is confusing me a bit, but did you mean to delete `warp/bin/linuxkudu`
> > 
> > 
> > Now that I look at it, the Github UI does seem confusing. But I haven't deleted the binary. I replaced it with a newer version of it that I got from artifacts of the kudu repository.
> 
> Can you update the macos kudus as well, we should keep them in sync.

We've decided that since building `kudu`s for macos is a pain, we only do it for PRs to `master`. For `develop` it's fine to only update the linux ones

**Comment by murcake** [comment]
Continuing in #146

**Pull request closed without merging**